### PR TITLE
Migrate `cross-lang-lto` `run-make` test to rmake

### PR DESCRIPTION
**Pull request by Oneirical**
Part of #121876 and the associated [Google Summer of Code project](https://blog.rust-lang.org/2024/05/01/gsoc-2024-selected-projects.html).

Please try:

try-job: x86_64-msvc
try-job: i686-mingw
try-job: x86_64-mingw
try-job: armhf-gnu
try-job: test-various
try-job: aarch64-apple
try-job: x86_64-gnu-llvm-18

### CODE REVIEW

File: tests/run-make/cross-lang-lto/rmake.rs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        emit_obj: true,
+    });
+}
+
+#[track_caller]
+fn check_bitcode(instructions: LibBuild) {
+    let mut rustc = rustc();
+    rustc
+        .input(instructions.source)
+        .output(instructions.output.clone())
```
Comment (jieyouxu) [inline tests/run-make/cross-lang-lto/rmake.rs:74]: Nit: can these clones be `&instructions.output`?

File: tests/run-make/cross-lang-lto/rmake.rs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+};
+
+fn main() {
+    check_bitcode(LibBuild {
+        source: path("lib.rs"),
+        crate_type: Some("staticlib"),
+        output: path(static_lib_name("liblib")),
+        lto: None,
+        emit_obj: false,
+    });
```
Comment (jieyouxu) [inline tests/run-make/cross-lang-lto/rmake.rs:24]: Remark: I like this API!

File: tests/run-make/cross-lang-lto/rmake.rs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    }
+    if let Some(crate_type) = instructions.crate_type {
+        rustc.crate_type(crate_type);
+    }
+    if let Some(lto) = instructions.lto {
+        rustc.arg(format!("-Clto={lto}"));
+    }
+    rustc.run();
+
+    if instructions.output.clone().extension().unwrap().to_str().unwrap() != "o" {
```
Comment (jieyouxu) [inline tests/run-make/cross-lang-lto/rmake.rs:89]: Suggestion:

```suggestion
    if instructions.output.clone().extension().unwrap() != "o" {
```

(`OsStr` has `PartialEq` impl with `&str` IIRC)

File: tests/run-make/cross-lang-lto/rmake.rs [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        // Remove all potential leftover object files, then turn the output into an object file.
+        for object in shallow_find_files(cwd(), |path| has_extension(path, "o")) {
+            rfs::remove_file(object);
+        }
+        llvm_ar().arg("x").arg(instructions.output.clone()).run();
```
Comment (jieyouxu) [inline tests/run-make/cross-lang-lto/rmake.rs:94]: Discussion: should we add a helper for `llvm-ar` for what `x` means?
Nit: `.arg(&instructions.output)`

**Review by jieyouxu — APPROVED** [review]
Thanks, changes look good, just some minor nits.

**Comment by Oneirical** [comment]
All changes applied.

Sorry, I didn't see you started a try-job, not sure if my push cancelled it.

  @rustbot ready

**Comment by jieyouxu** [comment]
> All changes applied.
> 
> Sorry, I didn't see you started a try-job, not sure if my push cancelled it.

No it's fine, AFAIK the try job will continue on the commit I started it on (only cosmetic changes, so no worries).

**Review by jieyouxu — APPROVED** [review]
Cool 😎

**Comment by jieyouxu** [comment]
The try job actually succeeded, it just didn't post a message.

@bors r+ rollup=iffy (cross language LTO)

**Comment by jieyouxu** [comment]
r=me with a rebase

@bors delegate+

**Pull request merged**